src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -80,49 +83,54 @@ public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Cancellati
+
                     return GetDiagnosticData();
                 }
 
-                await AppendDiagnosticsAsync(Solution, cancellationToken).ConfigureAwait(false);
+                await ProduceSolutionDiagnosticsAsync(Solution, cancellationToken).ConfigureAwait(false);
                 return GetDiagnosticData();
             }
 
-            protected async Task AppendDiagnosticsAsync(Solution solution, CancellationToken cancellationToken)
+            protected Task ProduceSolutionDiagnosticsAsync(Solution solution, CancellationToken cancellationToken)
+                => ProduceProjectDiagnosticsAsync(solution.Projects, static project => project.DocumentIds, includeProjectNonLocalResult: true, cancellationToken);
+
+            protected async Task ProduceProjectDiagnosticsAsync(
+                IEnumerable<Project> projects, Func<Project, IReadOnlyList<DocumentId>> getDocumentIds,
+                bool includeProjectNonLocalResult, CancellationToken cancellationToken)
             {
                 // PERF: run projects in parallel rather than running CompilationWithAnalyzer with concurrency == true.
                 // We do this to not get into thread starvation causing hundreds of threads to be spawned.
-                var includeProjectNonLocalResult = true;
-
-                var tasks = new Task[solution.ProjectIds.Count];
-                var index = 0;
-                foreach (var project in solution.Projects)
-                {
-                    var localProject = project;
-                    tasks[index++] = Task.Run(
-                        () => AppendDiagnosticsAsync(
-                            localProject, localProject.DocumentIds, includeProjectNonLocalResult, cancellationToken), cancellationToken);
-                }
-
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                var diagnostics = await ProducerConsumer<DiagnosticData>.RunParallelAsync(
+                    source: projects,
+                    produceItems: static (project, callback, args, cancellationToken) => args.@this.ProduceDiagnosticsAsync(
+                        project, args.getDocumentIds(project), args.includeProjectNonLocalResult, callback, cancellationToken),
+                    args: (@this: this, getDocumentIds, includeProjectNonLocalResult),
+                    cancellationToken).ConfigureAwait(false);
+
+                AppendDiagnostics(diagnostics);
             }
 
             protected void AppendDiagnostics(ImmutableArray<DiagnosticData> items)
             {
                 Debug.Assert(!items.IsDefault);
 
                 if (_lazyDataBuilder == null)
-                {
                     Interlocked.CompareExchange(ref _lazyDataBuilder, ImmutableArray.CreateBuilder<DiagnosticData>(), null);
-                }
 
                 lock (_lazyDataBuilder)
```

Comment:
>lock (_lazyDataBuilder) [](http://example.com/codeflow?start=16&length=23)

Curious whether this lock is still needed (can GetDiagnosticsAsync get called multiple times on the same instance of an object?)

Reply:
removed.